python/tvm/relay/transform/quantization/quantized_operators/dense.py

```diff
@@ -0,0 +1,183 @@
+
+    # Assume this casting is a no-op in the case we are casting back to itself
+    weight_zero_point = relay.cast(weight_zero_point, internal_accumulation_dtype)
+    data_zero_point = relay.cast(data_zero_point, internal_accumulation_dtype)
+
+    data = relay.cast(data, internal_accumulation_dtype)
+    weight = relay.cast(weight, internal_accumulation_dtype)
+
+    first_term = nn.dense(data, weight, units=out_units, out_dtype=internal_accumulation_dtype)
+
+    # The fields for the reduction operations to make things clear
+    axis = [1]
+    keep_dims = True
+    exclude = False
+
+    second_term = tensor._make.sum(data, axis, keep_dims, exclude) * weight_zero_point
+
+    # The fields for the reduction operations to make things clear
+    axis = [1]
+    keep_dims = False
+    exclude = False
+    third_term = tensor._make.sum(weight, axis, keep_dims, exclude) * data_zero_point
+
+    fourth_term = (
+        relay.Constant(tvm.nd.array(np.array(in_units, dtype=internal_accumulation_dtype)))
+        * data_zero_point
+        * weight_zero_point
+    )
+
+    # TODO: simulate overflow for other data types
+
+    # return first_term, second_term, third_term, fourth_term
+    return (first_term - second_term - third_term + fourth_term), QParams(
+        data_scale * weight_scale, 0, simulated_accumulation_dtype
+    )
+
+
+def generate_static_quantized_dense(
+    data: tvm.relay.Expr,
+    weight: tvm.relay.Expr,
+    data_qparams: QParams,
+    weight_qparams: QParams,
+    accumulation_dtype: str = "int32",
+    out_units: Optional[int] = None,
+    in_units: Optional[int] = None,
+) -> Tuple[tvm.relay.Expr, QParams]:
+    return generate_generic_quantized_dense(
+        data,
+        weight,
+        data_qparams,
+        weight_qparams,
+        internal_accumulation_dtype=accumulation_dtype,
+        simulated_accumulation_dtype=accumulation_dtype,
+        out_units=out_units,
+        in_units=in_units,
+    )
+
+
+def generate_simulated_quantized_dense(
+    data: tvm.relay.Expr,
+    weight: tvm.relay.Expr,
+    data_qparams: QParams,
+    weight_qparams: QParams,
+    simulated_accumulation_dtype: str = "int32",
+    out_units: Optional[int] = None,
+    in_units: Optional[int] = None,
+) -> Tuple[tvm.relay.Expr, QParams]:
+    return generate_generic_quantized_dense(
+        data,
+        weight,
+        data_qparams,
+        weight_qparams,
+        internal_accumulation_dtype="float32",
+        simulated_accumulation_dtype=simulated_accumulation_dtype,
+        out_units=out_units,
+        in_units=in_units,
+    )
+
+
+def example_dense_simulated(n, in_units, out_units, seed=42):
+    np.random.seed(seed=seed)
+    data_arr = np.random.randint(-10, 10, size=(n, in_units)).astype("float32")
+    weight_arr = np.random.randint(-10, 10, size=(out_units, in_units)).astype("float32")
+
+    var_creator = AffineQuantizationVarCreator()
+    data = relay.Var("data")
```

Comment:
Here, you probably should use `relay.var` instead of `relay.Var`. The ones that are capitalized are class constructors, whereas the lower-case ones are supposed to be user-facing. Also, you'll need to pass in either a type annotation, or a shape and a dtype into `relay.var`. 

https://tvm.apache.org/docs/api/python/relay/index.html?highlight=relay%20var#tvm.relay.var

Reply:
Actually, it doesn't really matter, since you're not putting a type in. relay.var is just syntactic sugar for relay.Var, and it'll construct the type for you automatically if you pass one in. But since you're not putting a type in, it doesn't really matter.

Reply:
I've added a type annotations for them at the least and used relay.var.

I now have learned always use the lower case ops since they are more pythonic and cast things for you :)